### PR TITLE
add instrumentation for intermittent DNS failures

### PR DESCRIPTION
**Pull request by wfurt**
per discussion in #34187, this PR adds new exception so we can track if retry would make improvements for TryAgain error cases.

### CODE REVIEW

File: src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs [hunk 1]
```diff
@@ -44,12 +43,21 @@ public void TryGetAddrInfo_HostName()
...
             {
                 // On Unix, we are not guaranteed to be able to resove the local host. The ability to do so depends on the 
                 // machine configurations, which varies by distro and is often inconsistent.
                 return;
             }
+
+            // Temporary instrumentation for #32797
+            if (error == SocketError.TryAgain && (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX) || RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD)))
+            {
+                if (error != SocketError.TryAgain)
```
Comment (stephentoub) [inline src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs]: How could this be true?

File: src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs [hunk 2]
```diff
@@ -105,10 +112,19 @@ public void TryGetAddrInfo_HostName_TryGetNameInfo()
...
                 // is turned off. Hence dns lookup for it's own hostname fails.
-                Assert.True(RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX));
+                Assert.True(RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX) || RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD));
                 return;
             }
 
+            // Temporary instrumentation for #32797
+            if (error == SocketError.TryAgain && (RuntimeInformation.IsOSPlatform(OSPlatform.Linux) || RuntimeInformation.IsOSPlatform(OSPlatform.OSX) || RuntimeInformation.IsOSPlatform(OSPlatform.FreeBSD)))
+            {
+                if (error != SocketError.TryAgain)
```
Comment (stephentoub) [inline src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs]: Same question... did you mean to retry?

**Comment by wfurt** [comment]
note that I saw one failure when running on my Ubuntu system attached to corp net:

```
   Starting:    System.Net.NameResolution.Pal.Tests (parallel test collections = on, max threads = 8)
      System.Net.NameResolution.PalTests.NameResolutionPalTests.TryGetAddrInfo_ExternalHost [FAIL]
        Assert.Equal() Failure
        Expected: Success
        Actual:   HostNotFound
        Stack Trace:
          /home/furt/github/wfurt-corefx-arm/src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs(146,0): at System.Net.NameResolution.PalTests.NameResolutionPalTests.TryGetAddrInfo_ExternalHost()
    Finished:    System.Net.NameResolution.Pal.Tests
  === TEST EXECUTION SUMMARY ===
     System.Net.NameResolution.Pal.Tests  Total: 11, Errors: 0, Failed: 1, Skipped: 0, Time: 0.348s
  ~/github/wfurt-corefx-arm/src/System.Net.NameResolution/tests/PalTests
```
when I run it again it passed. This is annoying and more difficult to handle as it returns permanent failure claiming that microsoft.com does not exist.

File: src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs [hunk 3]
```diff
@@ -35,7 +35,6 @@ public void TryGetAddrInfo_LocalHost()
             Assert.NotNull(hostEntry.Aliases);
         }
 
-        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotArm64Process))] // [ActiveIssue(32797)]
```
Comment (caesar1995) [inline src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs]: This is a test case inside a test project, we need [Fact]/[Theory] label to run the test. Here seems we remove the [ConditionalFact] without adding a new [Fact] label, and I don't think the test will actually run - or this is intentional? not running the test?

cc: @davidsh

Comment (davidsh) [reply src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs]: I agree.  And there is another test in this file also that now no longer runs.  It needs to be changed to "[Fact]" as well.

File: src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs [hunk 4]
```diff
@@ -50,6 +49,16 @@ public void TryGetAddrInfo_HostName()
...
+            if (error == SocketError.TryAgain && Environment.OSVersion.Platform == PlatformID.Unix)
+            {
+                error = NameResolutionPal.TryGetAddrInfo(hostName, out hostEntry, out nativeErrorCode);
+                if (error != SocketError.TryAgain)
+                {
+                    throw new InvalidOperationException("Name resolution failure preventable with retry");
+                }
+            }
             
             Assert.Equal(SocketError.Success, error);
```
Comment (stephentoub) [inline src/System.Net.NameResolution/tests/PalTests/NameResolutionPalTests.cs]: The test just failed again at this assert, meaning the retry didn't help, at least not in that case:
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F34956~2Fmerge/test~2Ffunctional~2Fcli~2F/20190130.31/workItem/System.Net.NameResolution.Pal.Tests/analysis/xunit/System.Net.NameResolution.PalTests.NameResolutionPalTests~2FTryGetAddrInfo_HostName

**Pull request merged**